ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/ContractKeysValidation.scala

```diff
@@ -66,27 +67,23 @@ private[transaction] object ContractKeysValidation {
       transactionEntry: DamlTransactionEntrySummary,
   ): StepResult[DamlTransactionEntrySummary] = {
     val keysValidationOutcome = transactionEntry.transaction
-      .foldInExecutionOrder[KeyValidationStatus](
-        Right(
-          KeyValidationState(activeStateKeys = contractKeyDamlStateKeys)
-        )
+      .foldInExecutionOrder[Either[KeyValidationError, List[KeyValidationState]]](
+        Right(List(KeyValidationState(activeStateKeys = contractKeyDamlStateKeys)))
       )(
-        (keyValidationStatus, _, exerciseBeginNode) =>
-          (
-            checkNodeContractKey(
-              exerciseBeginNode,
-              contractKeysToContractIds,
-              keyValidationStatus,
-            ),
-            true,
-          ),
-        (_, _, _) =>
-          // TODO https://github.com/digital-asset/daml/issues/8020
-          sys.error("rollback nodes are not supported"),
-        (keyValidationStatus, _, leafNode) =>
-          checkNodeContractKey(leafNode, contractKeysToContractIds, keyValidationStatus),
-        (accum, _, _) => accum,
-        (accum, _, _) => accum,
+        exerciseBegin = (status, _, exerciseBeginNode) => {
+          val newStatus =
+            onCurrentStatus(status)(
+              checkNodeContractKey(exerciseBeginNode, contractKeysToContractIds, _)
+            )
+          (newStatus, true)
+        },
+        rollbackBegin = (status, _, _) =>
+          // Store state to restore after rollback
+          (status.map(stack => stack.head +: stack), true),
```

Comment:
But anyway, aren't rollback blocks flattened?

Reply:
No, there is some level of normalization but in general we cannot flatten completely. Consider something like
rb - create
     - rb - create
      - exercise
If the exercise here is on the first create then it is valid. But if the exercise is on the second create then it has been rolled back and should fail.